### ISSUE TITLE
Strict mode Pyright complains about certain types requiring generic arguments, however, these types are not generics in Python 3.8

### ISSUE THREAD

**Issue by WSH032** [microsoft/pyright#7893]
**Describe the bug**

Strict mode Pyright complains about certain types requiring generic arguments, however, these types are not generics in Python 3.8

At least for `AbstractContextManager` and `Popen`, there are such issues. Other standard library types may also have similar problems, but I haven't tested them.

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.8&strict=true&locale=en-us&code=GYJw9gtgBAxmB2AXApgD0QGwJYCMpYgAcwREoBBHAZ0RAEMZEBhBFdAWTvjoHNkQAUKEhQqAVxyFwMZFSr4iJMgAUwhZPAFaAxFACiqdY2QATKIgCe6qHRA8xEDYnnASUPvH5YYsDHTlQAESUNPSMLEhoiJzcfCCBAjB%2BAQBCtgAUIbQMzKxRMbz8AJQAXAJQFVCE-lQ6%2BobIxmaW1rb2jkgubh5ePkk1Qarq8An9qXQAXulDGqXlldVyAkA)

```python
from contextlib import AbstractContextManager
from subprocess import Popen


# Expected type arguments for generic class "AbstractContextManager"
class Bar(AbstractContextManager):
    pass

# Expected type arguments for generic class "Popen"
class Baz(Popen):
    pass

```

---

Providing generic arguments indeed make pyright happy, however, running this code in Python 3.8 will result in errors.

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.8&strict=true&locale=en-us&code=GYJw9gtgBAxmB2AXApgD0QGwJYCMpYgAcwREoBBHAZ0RAEMZEBhBFdAWTvjoHNkQAUKEhQqAVxyFwMZFSr4iJMgAUwhZPAFaAxFAAqAT3UBREOBAAuKAHJyAISbtkiOtahgcAK2SN88%2BGBk4tQwIFiELjgYyAIwGHRyUHZ0IAAUlDT0jCxIaIic3HwgANoARMkgpQC6AJQWAlCNUIQJVDr6Rsim5lbWiJ1uHt6%2BWP6BohJUoeGR0bHxickAXqmq6vDFmbX1Tc2tAkA)

```python
from contextlib import AbstractContextManager
from subprocess import Popen


# TypeError: 'ABCMeta' object is not subscriptable
class Bar(AbstractContextManager["Bar"]):
    pass

# TypeError: 'type' object is not subscriptable
class Baz(Popen[str]):
    pass

```

**Related issues**

https://github.com/microsoft/pyright/issues/2246#issuecomment-909750736

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

Pyright works from type information provided by the typeshed project. It indicates that [AbstractContextManager](https://github.com/python/typeshed/blob/6565e8a0a0efbd8735a473e6c0d57f787c7fa4a2/stdlib/contextlib.pyi#L42) and [Popen](https://github.com/python/typeshed/blob/6565e8a0a0efbd8735a473e6c0d57f787c7fa4a2/stdlib/subprocess.pyi#L1837) are generic classes. If you think this is an error, you could file a bug report in the typeshed project.

If you're able to do so, consider upgrading to a version of python newer than 3.8, which is quite old at this point and will no longer be "in support" later this year.

If you are not yet able to move to a more up-to-date version of python, you will likely need to disable strict mode type checking or suppress individual errors.

**Comment by WSH032** [comment]
Thank you for your quick response! 👍 

The upgrade is not a problem for me personally, but I maintain some libraries that need to support 3.8, after all, 3.8 is still within its lifecycle.

Although it might make CI testing a bit awkward, I can accept that.

**Comment by hmc-cs-mdrissi** [comment]
My typical solution for this issue as I also mostly work on older python is,

```python
if TYPE_CHECKING:
  BazBase = Popen[str]
else:
  BazBase = POpen

class Baz(BazBase):
  ...

**Comment by WSH032** [comment]
Thanks, @hmc-cs-mdrissi ! That's exactly what I did.

But unfortunately, it doesn't work for `AbstractContextManager`

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.8&strict=true&locale=en-us&code=GYJw9gtgBAxmB2AXApgD0QGwJYCMpYgAcwREoBBHAZ0RAEMZEBhBFdAWTvjoHNkQAUKEhREAT0JZ4PfERJkAKgE0ACgFEA%2BkwASapgGkAkgDkA4gBoK8MQNtZgUZeq26DJ0wC4BUH1A2UaekYWJDRETm4%2BEAUJZCgAXgpqWgZmVjCI3n4AbQAiACE6EFyAXQFkDCpkL18-AJTg9I4uLOjYhKTA1JC2cJao2wBiChAeAFcIZCRRMFgMOioqKAgxmigcOLp1hbiYecWBYcKquYWl8UI4rCWx%2BABreDAAd3hLMGoYMZApGQu4sAcABN%2BFgAG7IQGnA57M5QQogAAU-mSQTSoWakX4MUuAEoar5gQ4NBo0FhEMSEVUMMBLAAqDQeKxiPHeWq%2BQhnIaOdoAqC5HBFXL4G73R4vARPMkACzhRQROKgC22IHxPhAyHBdAwGj%2BCIFIBxAiAA)

```python
from contextlib import AbstractContextManager
from typing import TYPE_CHECKING, Any


if TYPE_CHECKING:
    _AbstractContextManagerType = AbstractContextManager["Bar"]
else:
    _AbstractContextManagerType = AbstractContextManager

# Argument to class must be a base class
# Base class type is unknown, obscuring type of derived class
class Bar(_AbstractContextManagerType):
    def __exit__(self, *_: Any):
        pass

# Type of "bar" is unknown
with Bar() as bar:
    reveal_type(bar)

```

**Comment by hmc-cs-mdrissi** [comment]
Add : TypeAlias. That fixes the error.

```python

from typing_extensions import TypeAlias

if TYPE_CHECKING:
    _AbstractContextManagerType: TypeAlias = AbstractContextManager["Bar"]
else:
    _AbstractContextManagerType = AbstractContextManager
```

Older pyright versions were a bit more lax here and didn't require TypeAlias and I forgot it.

**Comment by WSH032** [comment]
`Popen["Bar"]` should be wrong, the generic argument of `Popen` is [constrained to `AnyStr`](https://github.com/python/typeshed/blob/6565e8a0a0efbd8735a473e6c0d57f787c7fa4a2/stdlib/subprocess.pyi#L1837). It seems Pyright did not detect this error.

**Comment by WSH032** [comment]
Perhaps deriving from `AbstractContextManager` was not the best approach.

I initially considered doing so, but it turned out to be [unfeasible](https://github.com/microsoft/pyright/discussions/7019):

```py
from contextlib import AbstractContextManager
from typing_extensions import Self


class Foo(AbstractContextManager[Self]):
    def __exit__(self, *_: object) -> None:
        pass
```

---

May be following would be better:

```py
from contextlib import AbstractContextManager
from typing_extensions import Self, assert_type


class Foo:
    def __enter__(self) -> Self:
        return self

    def __exit__(self, *_: object) -> None:
        pass


class Bar(Foo):
    pass


def func(bar: "AbstractContextManager[Bar]") -> None:
    assert isinstance(bar, AbstractContextManager)
    with bar as b:
        assert_type(b, Bar)


func(Bar())

```
